platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java

```diff
@@ -265,535 +335,134 @@ public Hash hash(
-                : "Improper computation of reservedLastLeafCount";
+        final int chunkHeight = config.virtualHasherChunkHeight();
+        int firstLeafRank = Path.getRank(firstLeafPath);
+        int lastLeafRank = Path.getRank(lastLeafPath);
 
         // Let the listener know we have started hashing.
         listener.onHashingStarted();
 
-        // Iterate over all dirty leaves until we encounter the very last segment (the so-called
-        // reserved segment).
-        long lastPath = -1;
-        while (itr.hasNext()) {
-            final VirtualLeafRecord<K, V> next = itr.peek();
-            final long path = next.getPath();
-            final int rank = getRank(path);
-
-            // SANITY CHECK: Fail fast if this condition does not hold
-            if (path < lastPath) {
-                throw new IllegalStateException("The paths in the iterator must be strictly increasing! " + "lastPath="
-                        + lastPath + ", path=" + path);
-            }
-            lastPath = path;
-
-            // The path must always be within this range. We can use an assertion here because,
-            // unless there is some bug in our code, this cannot happen.
-            assert path >= firstLeafPath && path <= lastLeafPath
-                    : "Invalid path lies outside the leaf path range " + path;
-
-            // Break out of the loop if the next leaf to process is in the reserved space
-            if (rank == maxLeafRank && path >= lastLeafPath - reservedLastLeafCount) {
-                break;
-            }
-
-            // Depending on the rank we're processing, we need to know the segment size.
-            final long segmentSize = rank == minLeafRank ? minRankSegmentSize : maxRankSegmentSize;
-            // Either eof is the end of the rank, or the last leaf before the reserved section
-            final long eofPath = Math.min((1L << (rank + 1)) - 1, lastLeafPath - reservedLastLeafCount + 1);
-
-            // Populate the wq with leaves by accumulating them from the iterator, a segment at a time.
-            final HashingQueue<K, V> wq = queue1.reset();
-            accumulate(itr, wq, path - (getIndexInRank(path) % segmentSize), segmentSize, eofPath);
-
-            // Setup and hash the subtree that we have accumulated. If I am hashing the minLeafRank,
-            // then the results go into the minRankStopQueue. If I am hashing the maxLeafRank, then
-            // the results go into the maxRankStopQueue.
-            final HashingQueue<K, V> pq = queue2.reset();
-            final HashingQueue<K, V> sq = rank == maxLeafRank ? maxRankStopQueue : minRankStopQueue;
-            listener.onBatchStarted();
-            hashSubTree(hashReader, listener, wq, pq, null, sq, firstLeafPath, lastLeafPath, rank, stopRank);
-            listener.onBatchCompleted();
+        final HashMap<Long, ChunkHashTask> map = new HashMap<>();
```

Comment:
We can try to run this code as a task in the pool ensuring all created tasks will be run in the pool as well.
Currently, profiling shows that we `execute()` many tasks created here and that is expensive. By making the current thread to be part of the pool we'll effectively use `fork()` instead.

Reply:
As we discussed offline, we will get this for free, once merkle tree hashing is changed to task based with its own fork-join pool.